config/eventvalidation.go

```diff
@@ -0,0 +1,140 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	validator "github.com/asaskevich/govalidator"
+	"github.com/golang/glog"
+)
+
+// VASTEventElement indicates valid VAST event element
+type VASTEventElement string
+
+const (
+	ImpressionVASTElement             VASTEventElement = "impression"
+	TrackingVASTElement               VASTEventElement = "tracking"
+	ClickTrackingVASTElement          VASTEventElement = "clicktracking"
+	CompanionClickThroughVASTElement  VASTEventElement = "companionclickthrough"
+	ErrorVASTElement                  VASTEventElement = "error"
+	NonLinearClickTrackingVASTElement VASTEventElement = "nonlinearclicktracking"
+)
+
+var vastEventElementMap = map[VASTEventElement]struct{}{
+	ImpressionVASTElement:             {},
+	TrackingVASTElement:               {},
+	ClickTrackingVASTElement:          {},
+	CompanionClickThroughVASTElement:  {},
+	ErrorVASTElement:                  {},
+	NonLinearClickTrackingVASTElement: {},
+}
+
+// TrackingEventType indicates quartile events
+type TrackingEventType string
+
+const (
+	Start         TrackingEventType = "start"
+	FirstQuartile TrackingEventType = "firstQuartile"
+	MidPoint      TrackingEventType = "midPoint"
+	ThirdQuartile TrackingEventType = "thirdQuartile"
+	Complete      TrackingEventType = "complete"
+)
+
+var trackingEventTypeMap = map[TrackingEventType]struct{}{
+	Start:         {},
+	FirstQuartile: {},
+	MidPoint:      {},
+	ThirdQuartile: {},
+	Complete:      {},
+}
+
+// validate verifies the events object  and returns error if at least one is invalid.
+func (e Events) validate(errs []error) []error {
+	if e.Enabled { // validate only if events are enabled
+		glog.Warning(`Don't enable this feature. It is still under developmment - https://github.com/prebid/prebid-server/issues/1725`)
+		if !isValidURL(e.DefaultURL) {
+			return append(errs, errors.New("Invalid events.default_url"))
+		}
+		err := validateVASTEvents(e.VASTEvents)
+		if err != nil {
+			return append(errs, err)
+		}
+	}
+	return errs
+}
+
+// validateVASTEvents verifies the all VASTEvent objects and returns error if at least one is invalid.
+func validateVASTEvents(events []VASTEvent) error {
+	if events != nil {
+		for i, event := range events {
+			if err := validateVASTEvent(event, i); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// validateVASTEvent validates event object and  returns error if at least one is invalid
+func validateVASTEvent(event VASTEvent, index int) error {
```

Comment:
I'd recommend adding this function over the `VASTEvent` struct like you do for events
```
func (v VASTEvent) validate() error {}
```

And you probably don't need to pass in the index to this function. You can just return generic VAST event related errors from this function and append that error in the `validateVASTEvents` functions to another error string that mentions the index number of the VAST event. So, for example, for an invalid `create_element` value, you can just return `return fmt.Errorf("Not a valid create_element value")` on L82 and then on L72 you could do `return fmt.Errorf("Invalid events.vast_events[%d]. %w", i, err)`. What do you think?

Reply:
Added receiver for  `validateVASTEvent` .  Instead of passing index, now this function will form error using following similar statement
```go
 fmt.Errorf("events.vast_events[%s].type is not applicable for create element '%s'", "%d", e.CreateElement)
```
This formate will replace `vast_events[%s]` with `vast_events[%d]`, which in turn will be replaced with actual index by calling function.